### PR TITLE
enhancement(graphql api): Expose `events_in` & `events_out` metrics

### PR DESCRIPTION
**Pull request by ktff**
Ref. #6367

### CODE REVIEW

**Comment by ktff** [comment]
@leebenson could you review this?

**Comment by leebenson** [comment]
Thanks @ktff, this is on my radar. Currently working through the Tokio upgrade for #6610 then I'll take a look 👍🏻

**Comment by jszwedko** [comment]
@leebenson could you give this look?

File: src/api/schema/metrics/events_in.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    pub fn get_timestamp(&self) -> Option<DateTime<Utc>> {
+        self.0.data.timestamp
+    }
+
+    pub fn get_events_in_total(&self) -> f64 {
+        match self.0.data.value {
+            MetricValue::Counter { value } => value,
+            _ => 0.00,
```
Comment (jszwedko) [inline src/api/schema/metrics/events_in.rs:19]: I'm still a little concerned about conflating `0` for components that will never publish this metric. I could see users being surprised that it never changes. This concern exists with the existing `events_processed_total`s though, so I'm ok with following up on this.

**Review by jszwedko — APPROVED** [review]
This looks good to me, thanks @ktff !

File: src/api/schema/metrics/events_in.rs (hunk 1 above)
Comment (ktff) [reply src/api/schema/metrics/events_in.rs:19]: Follow up sounds good, and once we start using `events_in_*` we will be able to remove `events_processed_totals` which will make this simpler.

File: tests/api.rs [hunk 2]
```diff
@@ -495,6 +495,55 @@ mod tests {
         )
     }
 
+    #[test]
```
Comment (leebenson) [inline tests/api.rs:498]: I appreciate you adding tests (thanks!), but just noting here that `tests/api.rs` is currently commented out in master, due to an issue with forked test and the recent Tokio upgrade.

This is being tracked in #6984.

I think it's good to leave these in and just comment out the whole file, so when #6984 is closed, these new tests are inherited.

File: src/api/schema/metrics/transform/generic.rs [hunk 3]
```diff
@@ -24,4 +24,14 @@ impl GenericTransformMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current transform
```
Comment (leebenson) [inline src/api/schema/metrics/transform/generic.rs:28]:
```suggestion
    /// Total incoming events for the current transform
```

I think "incoming" and "outgoing" read a little easier than "inputted" and "outputted".

File: src/api/schema/metrics/transform/generic.rs [hunk 4]
```diff
@@ -24,4 +24,14 @@ impl GenericTransformMetrics {
...
+    pub async fn events_in_total(&self) -> Option<metrics::EventsInTotal> {
+        self.0.events_in_total()
+    }
+
+    /// Events outputted for the current transform
```
Comment (leebenson) [inline src/api/schema/metrics/transform/generic.rs:33]:
```suggestion
    /// Total outgoing events for the current transform
```

File: src/api/schema/metrics/source/generic.rs [hunk 5]
```diff
@@ -24,4 +24,14 @@ impl GenericSourceMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current source
```
Comment (leebenson) [inline src/api/schema/metrics/source/generic.rs:28]:
```suggestion
    /// Total incoming events for the current source
```

File: src/api/schema/metrics/source/generic.rs [hunk 6]
```diff
@@ -24,4 +24,14 @@ impl GenericSourceMetrics {
...
+    pub async fn events_in_total(&self) -> Option<metrics::EventsInTotal> {
+        self.0.events_in_total()
+    }
+
+    /// Events outputted for the current source
```
Comment (leebenson) [inline src/api/schema/metrics/source/generic.rs:33]:
```suggestion
    /// Total outgoing events for the current source
```

File: src/api/schema/metrics/source/file.rs [hunk 7]
```diff
@@ -160,6 +194,16 @@ impl FileSourceMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current file source
+    pub async fn events_in_total(&self) -> Option<metrics::EventsInTotal> {
+        self.0.events_in_total()
+    }
+
+    /// Events outputted for the current file source
```
Comment (leebenson) [inline src/api/schema/metrics/source/file.rs:203]:
```suggestion
    /// Total outgoing events for the current file source
```

File: src/api/schema/metrics/source/file.rs [hunk 8]
```diff
@@ -160,6 +194,16 @@ impl FileSourceMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current file source
```
Comment (leebenson) [inline src/api/schema/metrics/source/file.rs:198]:
```suggestion
    /// Total incoming events for the current file source
```

File: src/api/schema/metrics/source/file.rs [hunk 9]
```diff
@@ -43,6 +43,16 @@ impl<'a> FileSourceMetricFile<'a> {
     async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.metrics.processed_bytes_total()
     }
+
+    /// Metric indicating events inputted for the current file
+    async fn events_in_total(&self) -> Option<metrics::EventsInTotal> {
+        self.metrics.events_in_total()
+    }
+
+    /// Metric indicating events outputted for the current file
```
Comment (leebenson) [inline src/api/schema/metrics/source/file.rs:52]:
```suggestion
    /// Metric indicating outgoing events for the current file
```

File: src/api/schema/metrics/source/file.rs [hunk 10]
```diff
@@ -43,6 +43,16 @@ impl<'a> FileSourceMetricFile<'a> {
     async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.metrics.processed_bytes_total()
     }
+
+    /// Metric indicating events inputted for the current file
```
Comment (leebenson) [inline src/api/schema/metrics/source/file.rs:47]:
```suggestion
    /// Metric indicating incoming events for the current file
```

File: src/api/schema/metrics/sink/generic.rs [hunk 11]
```diff
@@ -24,4 +24,14 @@ impl GenericSinkMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current sink
+    pub async fn events_in_total(&self) -> Option<metrics::EventsInTotal> {
+        self.0.events_in_total()
+    }
+
+    /// Events outputted for the current sink
```
Comment (leebenson) [inline src/api/schema/metrics/sink/generic.rs:33]:
```suggestion
    /// Total outgoing events for the current sink
```

File: src/api/schema/metrics/sink/generic.rs [hunk 12]
```diff
@@ -24,4 +24,14 @@ impl GenericSinkMetrics {
     pub async fn processed_bytes_total(&self) -> Option<metrics::ProcessedBytesTotal> {
         self.0.processed_bytes_total()
     }
+
+    /// Events inputted for the current sink
```
Comment (leebenson) [inline src/api/schema/metrics/sink/generic.rs:28]:
```suggestion
    /// Total incoming events for the current sink
```

File: src/api/schema/metrics/mod.rs [hunk 13]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+                    ComponentEventsOutThroughput::new(
+                        m.tag_value("component_name").unwrap(),
+                        throughput as i64,
+                    )
+                })
+                .collect()
+        })
+    }
+
+    /// Component event outputting metrics over `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:202]:
```suggestion
    /// Total outgoing component event metrics over `interval`
```

Noting that we generally haven't added a period suffix to GraphQL doc comments so far. Removing for consistency.

File: src/api/schema/metrics/mod.rs [hunk 14]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+    /// Event outputting throughput sampled over the provided millisecond `interval`.
+    async fn events_out_throughput(
+        &self,
+        #[graphql(default = 1000, validator(IntRange(min = "10", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = i64> {
+        counter_throughput(interval, &|m| m.name() == "events_out_total")
+            .map(|(_, throughput)| throughput as i64)
+    }
+
+    /// Component event outputting throughput metrics over `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:185]:
```suggestion
    /// Total outgoing component event throughput metrics over `interval`
```

File: src/api/schema/metrics/mod.rs [hunk 15]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+        &self,
+        #[graphql(default = 1000, validator(IntRange(min = "10", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = EventsOutTotal> {
+        get_metrics(interval).filter_map(|m| match m.name() {
+            "events_out_total" => Some(EventsOutTotal::new(m)),
+            _ => None,
+        })
+    }
+
+    /// Event outputting throughput sampled over the provided millisecond `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:176]:
```suggestion
    /// Total outgoing events throughput sampled over the provided millisecond `interval`
```

File: src/api/schema/metrics/mod.rs [hunk 16]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+    /// Component event inputting metrics over `interval`.
+    async fn component_events_in_totals(
+        &self,
+        #[graphql(default = 1000, validator(IntRange(min = "10", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = Vec<ComponentEventsInTotal>> {
+        component_counter_metrics(interval, &|m| m.name() == "events_in_total")
+            .map(|m| m.into_iter().map(ComponentEventsInTotal::new).collect())
+    }
+
+    /// Event outputting metrics.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:165]:
```suggestion
    /// Total outgoing events metrics
```

File: src/api/schema/metrics/mod.rs [hunk 17]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+                    ComponentEventsInThroughput::new(
+                        m.tag_value("component_name").unwrap(),
+                        throughput as i64,
+                    )
+                })
+                .collect()
+        })
+    }
+
+    /// Component event inputting metrics over `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:156]:
```suggestion
    /// Total incoming component event metrics over `interval`
```

File: src/api/schema/metrics/mod.rs [hunk 18]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+    /// Event inputting throughput sampled over the provided millisecond `interval`.
+    async fn events_in_throughput(
+        &self,
+        #[graphql(default = 1000, validator(IntRange(min = "10", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = i64> {
+        counter_throughput(interval, &|m| m.name() == "events_in_total")
+            .map(|(_, throughput)| throughput as i64)
+    }
+
+    /// Component event inputting throughput metrics over `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:139]:
```suggestion
    /// Total incoming component events throughput metrics over `interval`
```

File: src/api/schema/metrics/mod.rs [hunk 19]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
...
+        &self,
+        #[graphql(default = 1000, validator(IntRange(min = "10", max = "60_000")))] interval: i32,
+    ) -> impl Stream<Item = EventsInTotal> {
+        get_metrics(interval).filter_map(|m| match m.name() {
+            "events_in_total" => Some(EventsInTotal::new(m)),
+            _ => None,
+        })
+    }
+
+    /// Event inputting throughput sampled over the provided millisecond `interval`.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:130]:
```suggestion
    /// Total incoming events throughput sampled over the provided millisecond `interval`
```

File: src/api/schema/metrics/mod.rs [hunk 20]
```diff
@@ -112,6 +116,98 @@ impl MetricsSubscription {
         })
     }
 
+    /// Event inputting metrics.
```
Comment (leebenson) [inline src/api/schema/metrics/mod.rs:119]:
```suggestion
    /// Total incoming events metrics
```

File: src/api/schema/metrics/events_out.rs [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+#[Object]
+impl ComponentEventsOutTotal {
+    /// Component name
+    async fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Events outputted total metric
```
Comment (leebenson) [inline src/api/schema/metrics/events_out.rs:67]:
```suggestion
    /// Total outgoing events metric
```

File: src/api/schema/metrics/events_out.rs [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+pub struct ComponentEventsOutTotal {
+    name: String,
+    metric: Metric,
+}
+
+impl ComponentEventsOutTotal {
+    /// Returns a new `ComponentEventsOutTotal` struct, which is a GraphQL type. The
+    /// component name is hoisted for clear field resolution in the resulting payload
```
Comment (leebenson) [inline src/api/schema/metrics/events_out.rs:50]:
```suggestion
    /// component name is hoisted for clear field resolution in the resulting payload.
```

File: src/api/schema/metrics/events_out.rs [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+#[Object]
+impl EventsOutTotal {
+    /// Metric timestamp
+    pub async fn timestamp(&self) -> Option<DateTime<Utc>> {
+        self.get_timestamp()
+    }
+
+    /// Total number of events outputted
```
Comment (leebenson) [inline src/api/schema/metrics/events_out.rs:31]:
```suggestion
    /// Total outgoing events
```

File: src/api/schema/metrics/events_in.rs [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+    }
+}
+
+pub struct ComponentEventsInThroughput {
+    name: String,
+    throughput: i64,
+}
+
+impl ComponentEventsInThroughput {
+    /// Returns a new `ComponentEventsInThroughput`, set to the provided name/throughput values
```
Comment (leebenson) [inline src/api/schema/metrics/events_in.rs:79]:
```suggestion
    /// Returns a new `ComponentEventsInThroughput`, set to the provided name/throughput values.
```

File: src/api/schema/metrics/events_in.rs [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+#[Object]
+impl ComponentEventsInTotal {
+    /// Component name
+    async fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Events inputted total metric
```
Comment (leebenson) [inline src/api/schema/metrics/events_in.rs:67]:
```suggestion
    /// Total incoming events metric
```

File: src/api/schema/metrics/events_in.rs [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+pub struct ComponentEventsInTotal {
+    name: String,
+    metric: Metric,
+}
+
+impl ComponentEventsInTotal {
+    /// Returns a new `ComponentEventsInTotal` struct, which is a GraphQL type. The
+    /// component name is hoisted for clear field resolution in the resulting payload
```
Comment (leebenson) [inline src/api/schema/metrics/events_in.rs:50]:
```suggestion
    /// component name is hoisted for clear field resolution in the resulting payload.
```

File: src/api/schema/metrics/events_in.rs [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+#[Object]
+impl EventsInTotal {
+    /// Metric timestamp
+    pub async fn timestamp(&self) -> Option<DateTime<Utc>> {
+        self.get_timestamp()
+    }
+
+    /// Total number of events inputted
```
Comment (leebenson) [inline src/api/schema/metrics/events_in.rs:31]:
```suggestion
    /// Total incoming events
```

File: src/api/schema/metrics/events_in.rs (hunk 1 above)
Comment (leebenson) [reply src/api/schema/metrics/events_in.rs:19]: Yeah, if we change this to return an `Option<f64>`, it should probably be done everywhere. I'm okay with leaving this in too for now to mimic current behaviour.

File: lib/vector-api-client/src/gql/metrics.rs [hunk 28]
```diff
@@ -222,4 +400,52 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+        interval: i64,
+    ) -> BoxedSubscription<ComponentEventsOutTotalsSubscription> {
+        let request_body = ComponentEventsOutTotalsSubscription::build_query(
+            component_events_out_totals_subscription::Variables { interval },
+        );
+
+        self.start::<ComponentEventsOutTotalsSubscription>(&request_body)
+    }
+
+    /// Executes an all component events out throughputs subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:440]:
```suggestion
    /// Executes an all component events out throughputs subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 29]
```diff
@@ -222,4 +400,52 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+        interval: i64,
+    ) -> BoxedSubscription<ComponentEventsInThroughputsSubscription> {
+        let request_body = ComponentEventsInThroughputsSubscription::build_query(
+            component_events_in_throughputs_subscription::Variables { interval },
+        );
+
+        self.start::<ComponentEventsInThroughputsSubscription>(&request_body)
+    }
+
+    /// Executes an all component events out totals subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:428]:
```suggestion
    /// Executes an all component events out totals subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 30]
```diff
@@ -222,4 +400,52 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+        interval: i64,
+    ) -> BoxedSubscription<ComponentEventsInTotalsSubscription> {
+        let request_body = ComponentEventsInTotalsSubscription::build_query(
+            component_events_in_totals_subscription::Variables { interval },
+        );
+
+        self.start::<ComponentEventsInTotalsSubscription>(&request_body)
+    }
+
+    /// Executes an all component events in throughputs subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:416]:
```suggestion
    /// Executes an all component events in throughputs subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 31]
```diff
@@ -222,4 +400,52 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
 
         self.start::<ComponentProcessedBytesThroughputsSubscription>(&request_body)
     }
+
+    /// Executes an all component events in totals subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:404]:
```suggestion
    /// Executes an all component events in totals subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 32]
```diff
@@ -175,6 +303,56 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+        interval: i64,
+    ) -> BoxedSubscription<EventsOutThroughputSubscription> {
+        let request_body = EventsOutThroughputSubscription::build_query(
+            events_out_throughput_subscription::Variables { interval },
+        );
+
+        self.start::<EventsOutThroughputSubscription>(&request_body)
+    }
+
     /// Executes an all component events processed totals subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:356]:
```suggestion
    /// Executes an all component events processed totals subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 33]
```diff
@@ -175,6 +303,56 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+    ) -> BoxedSubscription<EventsOutTotalSubscription> {
+        let request_body =
+            EventsOutTotalSubscription::build_query(events_out_total_subscription::Variables {
+                interval,
+            });
+
+        self.start::<EventsOutTotalSubscription>(&request_body)
+    }
+
+    /// Executes an events out throughput subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:344]:
```suggestion
    /// Executes an events out throughput subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 34]
```diff
@@ -175,6 +303,56 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+        interval: i64,
+    ) -> BoxedSubscription<EventsInThroughputSubscription> {
+        let request_body = EventsInThroughputSubscription::build_query(
+            events_in_throughput_subscription::Variables { interval },
+        );
+
+        self.start::<EventsInThroughputSubscription>(&request_body)
+    }
+
+    /// Executes an events out metrics subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:331]:
```suggestion
    /// Executes an events out metrics subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 35]
```diff
@@ -175,6 +303,56 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
...
+    ) -> BoxedSubscription<EventsInTotalSubscription> {
+        let request_body =
+            EventsInTotalSubscription::build_query(events_in_total_subscription::Variables {
+                interval,
+            });
+
+        self.start::<EventsInTotalSubscription>(&request_body)
+    }
+
+    /// Executes an events in throughput subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:319]:
```suggestion
    /// Executes an events in throughput subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 36]
```diff
@@ -175,6 +303,56 @@ impl MetricsSubscriptionExt for crate::SubscriptionClient {
         self.start::<ProcessedBytesThroughputSubscription>(&request_body)
     }
 
+    /// Executes an events in metrics subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:306]:
```suggestion
    /// Executes an events in metrics subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 37]
```diff
@@ -129,6 +233,30 @@ pub trait MetricsSubscriptionExt {
...
+        interval: i64,
+    ) -> crate::BoxedSubscription<ComponentEventsInThroughputsSubscription>;
+
+    /// Executes an component events out totals subscription
+    fn component_events_out_totals_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<ComponentEventsOutTotalsSubscription>;
+
+    /// Executes an component events in throughputs subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:255]:
```suggestion
    /// Executes an component events in throughputs subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 38]
```diff
@@ -129,6 +233,30 @@ pub trait MetricsSubscriptionExt {
...
+        interval: i64,
+    ) -> crate::BoxedSubscription<ComponentEventsInTotalsSubscription>;
+
+    /// Executes an component events in throughputs subscription
+    fn component_events_in_throughputs_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<ComponentEventsInThroughputsSubscription>;
+
+    /// Executes an component events out totals subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:249]:
```suggestion
    /// Executes an component events out totals subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 39]
```diff
@@ -129,6 +233,30 @@ pub trait MetricsSubscriptionExt {
...
         interval: i64,
     ) -> crate::BoxedSubscription<ComponentProcessedBytesThroughputsSubscription>;
+
+    /// Executes an component events in totals subscription
+    fn component_events_in_totals_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<ComponentEventsInTotalsSubscription>;
+
+    /// Executes an component events in throughputs subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:243]:
```suggestion
    /// Executes an component events in throughputs subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 40]
```diff
@@ -129,6 +233,30 @@ pub trait MetricsSubscriptionExt {
         &self,
         interval: i64,
     ) -> crate::BoxedSubscription<ComponentProcessedBytesThroughputsSubscription>;
+
+    /// Executes an component events in totals subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:237]:
```suggestion
    /// Executes an component events in totals subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 41]
```diff
@@ -106,6 +186,30 @@ pub trait MetricsSubscriptionExt {
...
+        interval: i64,
+    ) -> crate::BoxedSubscription<EventsInThroughputSubscription>;
+
+    /// Executes an events out metrics subscription
+    fn events_out_total_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<EventsOutTotalSubscription>;
+
+    /// Executes an events out throughput subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:207]:
```suggestion
    /// Executes an events out throughput subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 42]
```diff
@@ -106,6 +186,30 @@ pub trait MetricsSubscriptionExt {
...
+        interval: i64,
+    ) -> crate::BoxedSubscription<EventsInTotalSubscription>;
+
+    /// Executes an events in throughput subscription
+    fn events_in_throughput_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<EventsInThroughputSubscription>;
+
+    /// Executes an events out metrics subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:201]:
```suggestion
    /// Executes an events out metrics subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 43]
```diff
@@ -106,6 +186,30 @@ pub trait MetricsSubscriptionExt {
         interval: i64,
     ) -> crate::BoxedSubscription<ProcessedBytesThroughputSubscription>;
 
+    /// Executes an events in metrics subscription
+    fn events_in_total_subscription(
+        &self,
+        interval: i64,
+    ) -> crate::BoxedSubscription<EventsInTotalSubscription>;
+
+    /// Executes an events in throughput subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:195]:
```suggestion
    /// Executes an events in throughput subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 44]
```diff
@@ -83,6 +123,46 @@ pub struct ComponentProcessedBytesThroughputsSubscription;
...
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/component_events_out_totals.graphql",
+    response_derives = "Debug"
+)]
+pub struct ComponentEventsOutTotalsSubscription;
+
 /// Extension methods for metrics subscriptions
 pub trait MetricsSubscriptionExt {
     /// Executes an uptime metrics subscription
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:168]:
```suggestion
    /// Executes an uptime metrics subscription.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 45]
```diff
@@ -83,6 +123,46 @@ pub struct ComponentProcessedBytesThroughputsSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/component_events_out_throughputs.graphql",
+    response_derives = "Debug"
+)]
+pub struct ComponentEventsOutThroughputsSubscription;
+
+/// ComponentEventsOutTotalsSubscription contains metrics on the number of events
+/// that have been emitted by a Vector instance, against specific components
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:157]:
```suggestion
/// that have been emitted by a Vector instance, against specific components.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 46]
```diff
@@ -83,6 +123,46 @@ pub struct ComponentProcessedBytesThroughputsSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/component_events_in_totals.graphql",
+    response_derives = "Debug"
+)]
+pub struct ComponentEventsInTotalsSubscription;
+
+/// ComponentEventsOutThroughputsSubscription contains metrics on the number of events
+/// that have been emitted between `interval` samples, against specific components
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:147]:
```suggestion
/// that have been emitted between `interval` samples, against specific components.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 47]
```diff
@@ -83,6 +123,46 @@ pub struct ComponentProcessedBytesThroughputsSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/component_events_in_throughputs.graphql",
+    response_derives = "Debug"
+)]
+pub struct ComponentEventsInThroughputsSubscription;
+
+/// ComponentEventsInTotalsSubscription contains metrics on the number of events
+/// that have been accepted for processing by a Vector instance, against specific components
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:137]:
```suggestion
/// that have been accepted for processing by a Vector instance, against specific components.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 48]
```diff
@@ -83,6 +123,46 @@ pub struct ComponentProcessedBytesThroughputsSubscription;
 )]
 pub struct ComponentProcessedBytesTotalsSubscription;
 
+/// ComponentEventsInThroughputsSubscription contains metrics on the number of events
+/// that have been accepted for processing between `interval` samples, against specific components
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:127]:
```suggestion
/// that have been accepted for processing between `interval` samples, against specific components.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 49]
```diff
@@ -43,6 +43,46 @@ pub struct ProcessedEventsThroughputSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/events_out_total.graphql",
+    response_derives = "Debug"
+)]
+pub struct EventsOutTotalSubscription;
+
+/// EventsOutThroughputSubscription contains metrics on the number of events
+/// that have been emitted between `interval` samples
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:77]:
```suggestion
/// that have been emitted between `interval` samples.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 50]
```diff
@@ -43,6 +43,46 @@ pub struct ProcessedEventsThroughputSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/events_in_throughput.graphql",
+    response_derives = "Debug"
+)]
+pub struct EventsInThroughputSubscription;
+
+/// EventsOutTotalSubscription contains metrics on the number of events
+/// that have been emitted by a Vector instance
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:67]:
```suggestion
/// that have been emitted by a Vector instance.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 51]
```diff
@@ -43,6 +43,46 @@ pub struct ProcessedEventsThroughputSubscription;
...
+#[derive(GraphQLQuery, Debug, Copy, Clone)]
+#[graphql(
+    schema_path = "graphql/schema.json",
+    query_path = "graphql/subscriptions/events_in_total.graphql",
+    response_derives = "Debug"
+)]
+pub struct EventsInTotalSubscription;
+
+/// EventsInThroughputSubscription contains metrics on the number of events
+/// that have been accepted for processing between `interval` samples
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:57]:
```suggestion
/// that have been accepted for processing between `interval` samples.
```

File: lib/vector-api-client/src/gql/metrics.rs [hunk 52]
```diff
@@ -43,6 +43,46 @@ pub struct ProcessedEventsThroughputSubscription;
 )]
 pub struct ProcessedBytesThroughputSubscription;
 
+/// EventsInTotalSubscription contains metrics on the number of events
+/// that have been accepted for processing by a Vector instance
```
Comment (leebenson) [inline lib/vector-api-client/src/gql/metrics.rs:47]:
```suggestion
/// that have been accepted for processing by a Vector instance.
```

File: lib/vector-api-client/graphql/schema.json [hunk 53]
```diff
@@ -472,6 +472,92 @@
...
+                  "kind": "SCALAR",
+                  "name": "String",
+                  "ofType": null
+                }
+              }
+            },
+            {
+              "args": [],
+              "deprecationReason": null,
+              "description": "Events inputted throughput",
```
Comment (leebenson) [inline lib/vector-api-client/graphql/schema.json:498]: Just a reminder here that if the comments changes are accepted, a re-run of `make build-graphql-schema` will keep these in sync.

**Review by leebenson — CHANGES_REQUESTED** [review]
My apologies for taking so long to review this, @ktff.

This looks good, thanks for tackling! I think the only blocking change is ensuring `test/api.rs` is commented out, or tests are likely to continue failing in our macOS CI steps.

I've suggested a few doc comments changes to match a similar style to existing comments. This is extra nitpicky just because these form public schema docs.

It might also be worth rebasing to master to ensure this continues to work as expected against the Tokio upgrade. I don't see any obvious reason it wouldn't, but since this was such a big change I'd rather test this in branch.

**Comment by ktff** [comment]
@leebenson could you recheck this. There were no obvious changes after merging Tokio upgrade.

**Review by leebenson — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**